### PR TITLE
feat: integrate buy button

### LINKED ISSUE #4541: Gallery Item Desktop - integrate buy

### PR DESCRIPTION
**Pull request by daiagi**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring
- [x] Redesign 

## Context

- [x] Closes #4541 
- [ ] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=EfmnRhHaQqfT3phm4cUCHCU3gFVDoSPR1U9WXzMRQBMqZ4L)

#### Community participation

- [ ] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by daiagi** [comment]
TODO

match Buy button disabled style to figma

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (daiagi) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:67]: @roiLeo 
I'm not sure that this is the correct way to do this

more generally, I would like to know please....
lets say i'm looking to buy this item:
https://kodadot.xyz/bsx/gallery/2548799063-138
which cost 0.49 ksm on the BSX network
should my wallet contain 0.49 ksm or the BSX equivalent?

**Comment by yangwao** [comment]
what's about code?

**Comment by daiagi** [comment]
> what's about code?


What do you mean?

**Comment by yangwao** [comment]
> What do you mean?

Asking if the code is good to go :)

@Jarsen136 can have look 👀

File: components/gallery/GalleryItemAction/GalleryItemAction.vue
Comment (roiLeo) [inline components/gallery/GalleryItemAction/GalleryItemAction.vue:5]: could you explain this change? `nft.currentOwner`?

File: components/gallery/GalleryItemAction/GalleryItemAction.vue
Comment (roiLeo) [inline components/gallery/GalleryItemAction/GalleryItemAction.vue:7]:
```bash 
Type 'string | undefined' is not assignable to type 'string'.
Type 'undefined' is not assignable to type 'string'
```

File: components/gallery/GalleryItemAction/GalleryItemActionSlides.vue
Comment (roiLeo) [inline components/gallery/GalleryItemAction/GalleryItemActionSlides.vue:85]: shouldn't we have variables here?

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (roiLeo) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:103]: `useRmrkVersion` who's this?

Comment (roiLeo) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:103]: nvm I didn't saw composable, why do you need this for? `RmrkVersionMixin` can't be used?

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (roiLeo) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:73]: On bsx idk, it might have some transaction fees (but it's really minimal) whereas on RMRK we have integrated platform fees (around .3cent if i recall correctly `somePercentFromTX`)

**Review by roiLeo — COMMENTED** [review]
I could not test it on bsx because I'm blocked by #4574

File: components/gallery/GalleryItemAction/GalleryItemAction.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemAction.vue:5]: I don't think I wrote it.
I think it came from a merge with main

Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemAction.vue]: it was bad undefined props handling.
resolved

File: components/gallery/GalleryItemAction/GalleryItemAction.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemAction.vue]: resolved

File: components/gallery/GalleryItemAction/GalleryItemActionSlides.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemActionSlides.vue]: done

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:112]: mmm because this component is not a class component?

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:82]: the ```somePercentFromTx``` is used in ```getTranasactionParams``` function
I think this balance calaulation is correct, it does the same as galleryItem in bsx an in rmrek were doing respectively

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (roiLeo) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:112]: Aight Vue3 no mixin!

**Review by roiLeo — CHANGES_REQUESTED** [review]
✅ tested on snek

After purchasing:
- "BUY" button is still present & i'm not the currentOwner until I refresh the page (need to refetch data)
- missing `MessageNotify` component
- we could show "Activity" Tab after that to show user new event

After reload:
- "Make Offer" button is present while I'm the current owner 🐋

**Comment by daiagi** [comment]
> white_check_mark tested on snek
> 
> After purchasing:
> 
>     * "BUY" button is still present & i'm not the currentOwner until I refresh the page (need to refetch data)
> 
>     * missing `MessageNotify` component (ref #4576 in progress)
> 
>     * we could show "Activity" Tab after that to show user new event

blocked by this: 
image

same as #4574  ?

> 
> After reload:
> 
>     * "Make Offer" button is present while I'm the current owner whale2

fixed

**Comment by yangwao** [comment]
I guess once #4587 is closed or will have upcoming PR, which will be sometime in upcoming 12 hours I bet we can continue on this

**Comment by yangwao** [comment]
Once this is merged, you can update your `main` branch @daiagi 
- #4596 

I expect to be sometime soon

**Comment by daiagi** [comment]
> @daiagi updating branch to test things 

Cool.
I'll get back on this tomorrow

**Comment by daiagi** [comment]
>* "BUY" button is still present & i'm not the currentOwner until I refresh the page (need to refetch data)

@roiLeo  is there a  way to use ```$apollo.addSmartQuery``` as a composable?

 ```const { $apollo } = useNuxtApp()``` doesn't expose that method

ideally i would like to add a pollInterval parameter to useGraphql

**Comment by yangwao** [comment]
> looking at the code i see it will be triggered only on minting new NFT.

we want to reuse component for showing with minting and buying, could be made in separate issue if it's not there yet, append `&message=congrats` at item to see

**Comment by daiagi** [comment]
> we want to reuse component for showing with minting and buying, could be made in separate issue if it's not there yet, append `&message=congrats` at item to see

is the congrats message suppose to disappear after a certain time?
if yes, how long?

**Comment by roiLeo** [comment]
> > * "BUY" button is still present & i'm not the currentOwner until I refresh the page (need to refetch data)
> 
> @roiLeo is there a way to use `$apollo.addSmartQuery` as a composable?
> 
> `const { $apollo } = useNuxtApp()` doesn't expose that method

Have you tried `useGraphql` composable? I think it's doable with that

> ideally i would like to add a pollInterval parameter to useGraphql

😱 why would you need that?

> > missing MessageNotify component (ref #4576 in progress)
> 
> Just tested buy on old design. didn't see that component show up, looking at the code i see it will be triggered only on minting new NFT. am i missing something?

Just copy `MessageNotify` component used in GalleryItem

**Comment by roiLeo** [comment]
> I thought it would be more elegant, instead of doing something like `setInterval` in `GallryItem`. but guess not

Wait what? why would you need to add a query with interval? can't this be done in async?

File: libs/ui/src/components/NeoButton/NeoButton.vue
Comment (roiLeo) [inline libs/ui/src/components/NeoButton/NeoButton.vue:119]: same stuff has been done in #4575, there may be a conflict here

**Comment by daiagi** [comment]
> > I thought it would be more elegant, instead of doing something like `setInterval` in `GallryItem`. but guess not
> 
> Wait what? why would you need to add a query with interval? can't this be done in async?

alright, maybe i got it figured out wrong.
my point is of course to re fetch data after NFT is bought in order to refresh the UI
but now that we're talking about it i realize that instead of polling i can just call ```useGalleryItem`` once again when buy is finshed successfully

File: components/gallery/GalleryItem.vue
Comment (roiLeo) [inline components/gallery/GalleryItem.vue]:
```suggestion
      :subtitle="$t('mint.successNewNfts')"/>
```

File: components/gallery/GalleryItem.vue
Comment (roiLeo) [inline components/gallery/GalleryItem.vue]: typo `showCongrtsMessgae`

File: components/gallery/GalleryItem.vue
Comment (roiLeo) [inline components/gallery/GalleryItem.vue:108]: `useAuth` composable

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (roiLeo) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:157]: use urlPrefix here?

Comment (roiLeo) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:157]: nvm this feature is only used on rmrk, we might want to have it on other chain

**Comment by roiLeo** [comment]
Still need to reload page:
- to see if I'm the currentOwner
- to see new activity
- to hide make offer & buy button

+ hover issue

recording


<details><summary>Browser metadata</summary>





```
Path:      /snek/gallery/659233203-4
Browser:   Firefox 109.0 on Mac OS 10.15
Viewport:  1920 x 995 @1x
Language:  en
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-4558--koda-nuxt.netlify.app%2Fsnek%2Fgallery%2F659233203-4&os=OS%20X&os_version=Catalina&browser=Firefox&browser_version=109.0&resolution=1920x995&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-4558--koda-nuxt.netlify.app/snek/gallery/659233203-4) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=63b68c2a41fc0b000bb11855&commentID=63b68fc91cd7007c5ed79720&resolution=resolved)

File: libs/ui/src/components/NeoButton/NeoButton.vue
Comment (daiagi) [reply libs/ui/src/components/NeoButton/NeoButton.vue:119]: it's a result of a merge conflict i resolved.
I'll have a closer look

Comment (daiagi) [reply libs/ui/src/components/NeoButton/NeoButton.vue:103]: css is ok now. quite sure

**Comment by roiLeo** [comment]
> ```
> const onNFTBought = () => {
>   activeTab.value = tabs.activity
>   showCongratsMessgae.value = true
>   setTimeout(() => {
>     showCongratsMessgae.value = false
>   }, 10000)
>   const {nft: tempNFT} = useGalleryItem()
>   watch (tempNFT, () => {
>     console.log(tempNFT.value)
>     if(tempNFT.value) {
>       nft.value = tempNFT.value
>     }
>   })
> 
> }
> ```

Please, don't use timeout.. It's already set in `MessageNotify` component

> this function is triggered by an event emitted on the success callback of `howAboutToExecute` so after the NFT has been bought
> 
> the NFT value returned by `useGalleyItem` show the previous owner (the one the NFT has just been bought from) so the UI doesn't update to reflect new owner
> 
> @roiLeo suggestions?

so if I understand well even after the success callback, the nft is not updated yet with a new owner? Have you tried to do it manually? I didn't pay attention to the console if the request was sending afterward.
Maybe [Nuxt3 doc](https://nuxt.com/docs/getting-started/data-fetching#refreshing-data) can help?

**Comment by yangwao** [comment]
> the nft is not updated yet with a new owner?

isn't something like subscription for event would help here? You should receive new event once block is finalized and loaded in indexer, so 3 blocks (block varies 7s) + 5s indexer drift

cc @vikiival

**Comment by daiagi** [comment]
> > the nft is not updated yet with a new owner?
> 
> isn't something like subscription for event would help here? You should receive new event once block is finalized and loaded in indexer, so 3 blocks (block varies 7s) + 5s indexer drift
> 
> cc @vikiival

yes , that's pretty much what i'm looking for here

do we have tools ready to handle this?

File: components/gallery/GalleryItem.vue
Comment (daiagi) [reply components/gallery/GalleryItem.vue]: really?

**Comment by daiagi** [comment]
btw @yangwao don't mean to be pushy, but gentle reminder that I'm trying to reach you on discord
maybe you like some other channel better?

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:157]: any change required here? resolve?

Comment (vikiival) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue:157]: yup please use URL prefix

**Comment by vikiival** [comment]
> > the nft is not updated yet with a new owner?
> 
> isn't something like subscription for event would help here? You should receive new event once block is finalized and loaded in indexer, so 3 blocks (block varies 7s) + 5s indexer drift
> 
> cc @vikiival

Gallery detail is fetching this minimal data every 2seconds imo, so as @yangwao said it is not instant

**Comment by daiagi** [comment]
this is how it's done in rmrk GalleryItem:

```vue
  public mounted() {
    // used to poll nft every second after component initialization in order to prevent double spending
    this.$apollo.addSmartQuery<{ nft }>('nft', {
      client: this.client,
      query: nftByIdMinimal,
      manual: true,
      variables: {
        id: this.id,
      },
      result: ({ data }) => {
        this.nft = {
          ...this.nft,
          ...data.nft,
        }
      },
      pollInterval: 1000,
    })

    if (this.message === 'congrats') {
      this.toast(this.message)
    }
  }
```
my initial thought was to replicate this in the new GalleryItem
thing is, the method ```addSmartQuery``` doesn't exist on the apollo client we use in Vue3 composition API

```const { $apollo } = useNuxtApp()```

me and @roiLeo  had discussion about this above, but I'm still stuck on it
@yangwao @vikiival any ideas?

**Comment by roiLeo** [comment]
Have you tried something like:
```ts
  const { data } = useGraphql({
    queryPrefix: 'subsquid',
    queryName: 'nftByIdMinimal',
    variables: {
      id: this.id,
    },
    options: {
      pollInterval: 1000,
    }
  })
```

**Comment by daiagi** [comment]
> Have you tried something like:
> 
> ```ts
>   const { data } = useGraphql({
>     queryPrefix: 'subsquid',
>     queryName: 'nftByIdMinimal',
>     variables: {
>       id: this.id,
>     },
>     options: {
>       pollInterval: 1000,
>     }
>   })
> ```

I have not.
Thank you

**Comment by daiagi** [comment]
@roiLeo  I got it working, used ```$apollo.watchQuery``` in ```useGraphQl``` if pollInterval option is passed
@vikiival maybe #4614 is closed / not needed ?

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue
Comment (daiagi) [reply components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemBuy.vue]: done

**Comment by vikiival** [comment]
> if pollInterval option is passed

That's the current implementation that I would like to avoid

**Comment by yangwao** [comment]
@daiagi maybe merge #4633 here?

**Comment by daiagi** [comment]
@Jarsen136 I've merged your branch here, but it's not working for me

UI doesn't update to reflect new owner
console.log from ```useSubscriptionGraphQl``` shows the new owner correctly
and doFetch is getting called
but the result from doFetch shows old owner


image

**Comment by Jarsen136** [comment]
> @Jarsen136 I've merged your branch here, but it's not working for me
> 
> UI doesn't update to reflect new owner console.log from `useSubscriptionGraphQl` shows the new owner correctly and doFetch is getting called but the result from doFetch shows old owner
> 

Oh sorry, I forget to push a commit to the branch. Pls try to merge https://github.com/kodadot/nft-gallery/pull/4633 and test again : )
image

**Comment by daiagi** [comment]
> Oh sorry, I forget to push a fix commit to the branch. Pls try to merge https://github.com/kodadot/nft-gallery/pull/4633 again and test : )

working now

**Comment by daiagi** [comment]
ready to merge?

File: utils/constants.ts
Comment (roiLeo) [inline utils/constants.ts:43]: wut? 008 isn't yet ready

Comment (daiagi) [reply utils/constants.ts:43]: yep. 
#4633  needs to be merged first and them I'll resolve conflicts

**Comment by yangwao** [comment]
- #4633 
has been merged @daiagi 😇

**Comment by daiagi** [comment]
@roiLeo @Jarsen136  @vikiival  ready for review

**Comment by daiagi** [comment]
disabled state looking good:
image

:heavy_check_mark:  tested on https://deploy-preview-4558--koda-nuxt.netlify.app/snek/gallery/659233203-5
* it takes a few moments for UI to change after buying, but as far as i can tell all is working properly

**Review by roiLeo — APPROVED** [review]
Tested on `/snek/gallery/185770742-2?redesign=true`

✅ works for me

**Comment by yangwao** [comment]
pay 110 usd

@daiagi let's resolve conflicts and we can merge it, finally buys 🤑

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $110 @ 27.51 USD/KSM ~ 3.999 $KSM
🧗 [EfmnRhHaQqfT3phm4cUCHCU3gFVDoSPR1U9WXzMRQBMqZ4L](https://kusama.subscan.io/account/EfmnRhHaQqfT3phm4cUCHCU3gFVDoSPR1U9WXzMRQBMqZ4L)
🔗 [0xd6628388b660391ba08b082f3abdc69e19e22b4355b0c84b3b0c5af2f4bfdacf](https://kusama.subscan.io/extrinsic/0xd6628388b660391ba08b082f3abdc69e19e22b4355b0c84b3b0c5af2f4bfdacf)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by vikiival** [comment]
@daiagi there is a conflict

**Comment by daiagi** [comment]
@vikiival conflict resolved

**Pull request merged**
